### PR TITLE
8332497: javac prints an AssertionError when annotation processing runs on program with module imports

### PR DESCRIPTION
**Pull request by Evemose**
Fix is pretty simple: visitModuleImport in com.sun.tools.javac.tree.TreeScanner has notbeen overriden, so defaulted to Visitor::visitModuleImport, which forwards to Visitor::visitTree, which is also not overriden, and, therefore, threw AssertionError. 

PS: Im not even sure how it worked before without crashing, seems like there is some intermidiate implementation between this TreeScanner and actual scanners because otherwise it should have resultedin compile error the moment it encounter module importin any visitor

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8332497: javac prints an AssertionError when annotation processing runs on program with module imports`

### Issue
 * [JDK-8332497](https://bugs.openjdk.org/browse/JDK-8332497): javac prints an AssertionError when annotation processing runs on program with module imports (**Bug** - P3)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19292/head:pull/19292` \
`$ git checkout pull/19292`

Update a local copy of the PR: \
`$ git checkout pull/19292` \
`$ git pull https://git.openjdk.org/jdk.git pull/19292/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19292`

View PR using the GUI difftool: \
`$ git pr show -t 19292`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19292.diff">https://git.openjdk.org/jdk/pull/19292.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19292#issuecomment-2119205543)

### CODE REVIEW

**Comment by liach** [comment]
I recommend adding a minimal test case for this bug, like the one from the JBS issue.

**Comment by Evemose** [comment]
@liach been done. It`s my first time writing javac tests so I am not sure I have done everything in the best way, but I`ve tried my best

**Comment by lahodaj** [comment]
The test is not wrong, but could be made shorter like here:
https://github.com/lahodaj/jdk/blob/8e4c9f4091f443766e5f009317661260154a33a8/test/langtools/tools/javac/ImportModule.java#L750

Also, the similar `com.sun.tools.javac.tree.TreeTranslator` could also be made consistent and handle module imports. Although I doubt there is an end-user way to see the difference (I am not aware about any whole-AST translation in javac; only per-class translations; for the internal scanner, it is also more common to scan only classes, I think).

**Comment by Evemose** [comment]
> The test is not wrong, but could be made shorter like here: https://github.com/lahodaj/jdk/blob/8e4c9f4091f443766e5f009317661260154a33a8/test/langtools/tools/javac/ImportModule.java#L750
> 
> Also, the similar `com.sun.tools.javac.tree.TreeTranslator` could also be made consistent and handle module imports. Although I doubt there is an end-user way to see the difference (I am not aware about any whole-AST translation in javac; only per-class translations; for the internal scanner, it is also more common to scan only classes, I think).

Thanks for suggestion. Is it ok to just out bug-related test into some pre-existing test class? I just saw that issue-related tests are all in separate folders so figured its how it should be. Also will whe processing run if there is no annotated elemnts at all? If i remeber correctly, it will not

File: test/langtools/tools/javac/annotations/8332497/T8332497.java
Comment (liach) [inline test/langtools/tools/javac/annotations/8332497/T8332497.java:1]: You will need a license header (see other test files)
Also, the name of the test should reflect the problem, like `testModuleImport/TestModuleImport.java`, naming by bug ID is legacy and not recommended.

**Comment by liach** [comment]
> Is it ok to just out bug-related test into some pre-existing test class? I just saw that issue-related tests are all in separate folders so figured its how it should be. 

Yes. If you look at git blame (available from Intellij Idea annotations) you will find those Txxxxxxxx tests are from long ago; they are legacy and no longer encouraged in newer tests, which usually group by their subjects.

File: test/langtools/tools/javac/annotations/ModuleImportProcessingTest.java
Comment (jddarcy) [inline test/langtools/tools/javac/annotations/ModuleImportProcessingTest.java:1]: Looks like this test would work better as a "native" annotation processing test.

I recommend look in the `test/langtools/tools/javac/processing` directory hierarchy, specifically for test classes that subclass `JavacTestingAbstractProcessor`. HTH

Comment (Evemose) [reply test/langtools/tools/javac/annotations/ModuleImportProcessingTest.java]: Thanks for the hint. I refactored the test, It looks much better now

**Comment by tstuefe** [comment]
Drive-by comment: I was curious about the "crashes" in the bug description, but this is no crash. It is an AssertionError. Could we please improve JBS issue title and PR title to say that?

**Comment by Evemose** [comment]
> Drive-by comment: I was curious about the "crashes" in the bug description, but this is no crash. It is an AssertionError. Could we please improve JBS issue title and PR title to say that?

I would be happy to, but I am not the one who created the issue. I filed mine report too, but I guess it has been discarded as duplicate. I was just the one to do the fix. @lahodaj I guess this request should be adressed to you

**Comment by lahodaj** [comment]
I've changed the issue title to: "javac prints an AssertionError when annotation processing runs on program with module imports".

OTOH, not sure what's wrong with "crash". It crashes (the program stops working) with an exception.

**Comment by tstuefe** [comment]
> I've changed the issue title to: "javac prints an AssertionError when annotation processing runs on program with module imports".
> 
> OTOH, not sure what's wrong with "crash". It crashes (the program stops working) with an exception.

Maybe its just me. When I read "crash", I think segfault, or core dump. A javac core'ing would be interesting, that's why I looked.

**Comment by liach** [comment]
I think most Java programmers call a program crash with a stacktrace a "crash" and hs_err_pids as a "VM crash".

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (Evemose) [reply test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:36]: I have tried to replace it with `@enablePreview`, but compilation in ci fails with:
/home/runner/work/jdk/jdk/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:24: error: module imports are a preview feature and are disabled by default.
import module java.base;
       ^
  (use --enable-preview to enable module imports)
1 error
Im not sure what I am missing. I have tried to also run with `--enable-preview --source ${jdk.version}`, and it also fails. I saw that `TestUnnamedVariableElement` uses both `@enablePreview~ and `--enable-preview --source ${jdk.version}` for some reason, but this approach also fails for me

File: test/langtools/tools/javac/4241573/ModuleImportProcessingTest.java
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/4241573/ModuleImportProcessingTest.java:1]: it seems like this file should be removed

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (liach) [inline test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:36]:
```suggestion
 * @compile JavacTestingAbstractProcessor.java ModuleImportProcessingTest.java
```
I think the `build` action doesn't support building dependencies with preview, so `@build ModuleImportProcessingTest` is the one that's failing here. You can revert all those `--enable-preview` for `@compile/process` down below.

Comment (Evemose) [reply test/langtools/tools/javac/processing/ModuleImportProcessingTest.java]: Thanks for the suggestion. Lets see how it works

**Comment by Evemose** [comment]
what I figured is that  -J-Dtest.enable.preview=true option is added to `@compile/process`. But even when `--enable-preview --source ${jdk.version}` is explictily specified, and is present in command line, for some reason, java.lang.UnsupportedClassVersionError: Preview features are not enabled for ModuleImportProcessingTest (class file version 67.65535). Try running with '--enable-preview' si still thrown. I guess it is some sor of a bug in a jtreg, I will file it later. For now, I will rollback to toolbox-based test, because it, at least, works fine

**Comment by Evemose** [comment]
Do I have to somehow request review or just wait untill someone eventually does it?

**Comment by liach** [comment]
> Do I have to somehow request review or just wait untill someone eventually does it?

I think @lahodaj would review as Jan's assigned to the JBS issue. This patch is ready to review so let's wait for him to review and sponsor (process might lapse to next Monday)

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:84]: looking good, just a minor comment, here we usually use: `Integer.toString(Runtime.version().feature())` to get the source version, and for indentation we use 4 spaces

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:90]: you don't need to call getOutputLines if you won't do anything with the output

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:46]: just `@run main ModuleImportProcessingTest`

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (Evemose) [reply test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:84]: Im not sure about which indentaion you are talking here. You mean I should use whitespaces instead of tabs, or that I should add spaces betweeen some elements of the code?

**Comment by vicente-romero-oracle** [comment]
ok this is what I would do, please feel free to use it, it is applicable on top of your current code. As you can see I have reformatted the test to align it better with the rest of similar tests in our code base. Also as Jan suggested I have added an equivalent method to TreeTranslator for completeness.

```
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
index 59a7457e6d0..63778fb42ff 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
@@ -131,6 +131,11 @@ public void visitImport(JCImport tree) {
         result = tree;
     }
 
+    public void visitModuleImport(JCModuleImport tree) {
+        tree.module = translate(tree.module);
+        result = tree;
+    }
+
     public void visitClassDef(JCClassDecl tree) {
         tree.mods = translate(tree.mods);
         tree.typarams = translateTypeParams(tree.typarams);
diff --git a/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java b/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
index 88c7974a26e..760ec30820e 100644
--- a/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
+++ b/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
@@ -21,7 +21,16 @@
  * questions.
  */
 
-import toolbox.*;
+/**
+ * @test
+ * @bug 8332497
+ * @summary javac prints an AssertionError when annotation processing runs on program with module imports
+ * @library /tools/lib
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ *          jdk.compiler/com.sun.tools.javac.main
+ * @build toolbox.JavacTask toolbox.ToolBox toolbox.Task
+ * @run main ModuleImportProcessingTest
+ */
 
 import javax.annotation.processing.AbstractProcessor;
 import javax.annotation.processing.RoundEnvironment;
@@ -32,40 +41,49 @@
 import java.nio.file.Paths;
 import java.util.Set;
 
-/**
- * @test
- * @bug 8332497
- * @summary error: javac prints an AssertionError when annotation processing runs on program with module imports
- * @library /tools/lib
- * @modules jdk.compiler/com.sun.tools.javac.api
- *          jdk.compiler/com.sun.tools.javac.main
- * @build toolbox.JavacTask toolbox.ToolBox toolbox.Task
- * @run main ModuleImportProcessingTest
- */
-public class ModuleImportProcessingTest {
-    final toolbox.ToolBox tb = new ToolBox();
-    final Path base = Paths.get(".");
-    final String processedSource = """
-        import module java.base;
-        import java.lang.annotation.*;
-        public class Main {
-          public static void main(String[] args) {
-            List.of();
-          }
-          @Ann
-          private void test() {}
-          @Retention(RetentionPolicy.RUNTIME)
-          @Target(ElementType.METHOD)
-          public @interface Ann {}
-        }
-        """;
+import toolbox.TestRunner;
+import toolbox.ToolBox;
+import toolbox.JavacTask;
+import toolbox.Task;
+
+public class ModuleImportProcessingTest extends TestRunner {
+    ToolBox tb = new ToolBox();
+
+    public ModuleImportProcessingTest() {
+        super(System.err);
+    }
+
+    protected void runTests() throws Exception {
+        runTests(m -> new Object[] { Paths.get(m.getName()) });
+    }
 
-    public static void main(String[] args) throws Exception {
-        new ModuleImportProcessingTest().test();
+    Path[] findJavaFiles(Path... paths) throws Exception {
+        return tb.findJavaFiles(paths);
     }
 
-    public void test() throws Exception {
-        tb.writeJavaFiles(base, processedSource);
+    public static void main(String... args) throws Exception {
+        new ModuleImportProcessingTest().runTests();
+    }
+
+    @Test
+    public void test(Path base) throws Exception {
+        Path src = base.resolve("src");
+        tb.writeJavaFiles(src,
+                """
+                import module java.base;
+                import java.lang.annotation.*;
+                public class Main {
+                    public static void main(String[] args) {
+                        List.of();
+                    }
+                    @Ann
+                    private void test() {}
+
+                    @Retention(RetentionPolicy.RUNTIME)
+                    @Target(ElementType.METHOD)
+                    public @interface Ann {}
+                }
+                """);
         new toolbox.JavacTask(tb)
                 .options(
                         "-processor", AP.class.getName(),
@@ -73,24 +91,15 @@ public void test() throws Exception {
                         "-source", Integer.toString(Runtime.version().feature()),
                         "-proc:only"
                 )
-                .outdir(base.toString())
-                .files(base.resolve("Main.java"))
-                .run(Task.Expect.SUCCESS)
-                .writeAll();
+                .files(findJavaFiles(src))
+                .run();
     }
 
     @SupportedAnnotationTypes("*")
     public static final class AP extends AbstractProcessor {
-
         @Override
         public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
             return false;
         }
-
-        @Override
-        public SourceVersion getSupportedSourceVersion() {
-            return SourceVersion.latest();
-        }
-
     }
-}
\ No newline at end of file
+}
```

**Comment by Evemose** [comment]
> ok this is what I would do, please feel free to use it, it is applicable on top of your current code. As you can see I have reformatted the test to align it better with the rest of similar tests in our code base. Also as Jan suggested I have added an equivalent method to TreeTranslator for completeness.
> 
> ```
> diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
> index 59a7457e6d0..63778fb42ff 100644
> --- a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
> +++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
> @@ -131,6 +131,11 @@ public void visitImport(JCImport tree) {
>          result = tree;
>      }
>  
> +    public void visitModuleImport(JCModuleImport tree) {
> +        tree.module = translate(tree.module);
> +        result = tree;
> +    }
> +
>      public void visitClassDef(JCClassDecl tree) {
>          tree.mods = translate(tree.mods);
>          tree.typarams = translateTypeParams(tree.typarams);
> diff --git a/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java b/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
> index 88c7974a26e..760ec30820e 100644
> --- a/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
> +++ b/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
> @@ -21,7 +21,16 @@
>   * questions.
>   */
>  
> -import toolbox.*;
> +/**
> + * @test
> + * @bug 8332497
> + * @summary javac prints an AssertionError when annotation processing runs on program with module imports
> + * @library /tools/lib
> + * @modules jdk.compiler/com.sun.tools.javac.api
> + *          jdk.compiler/com.sun.tools.javac.main
> + * @build toolbox.JavacTask toolbox.ToolBox toolbox.Task
> + * @run main ModuleImportProcessingTest
> + */
>  
>  import javax.annotation.processing.AbstractProcessor;
>  import javax.annotation.processing.RoundEnvironment;
> @@ -32,40 +41,49 @@
>  import java.nio.file.Paths;
>  import java.util.Set;
>  
> -/**
> - * @test
> - * @bug 8332497
> - * @summary error: javac prints an AssertionError when annotation processing runs on program with module imports
> - * @library /tools/lib
> - * @modules jdk.compiler/com.sun.tools.javac.api
> - *          jdk.compiler/com.sun.tools.javac.main
> - * @build toolbox.JavacTask toolbox.ToolBox toolbox.Task
> - * @run main ModuleImportProcessingTest
> - */
> -public class ModuleImportProcessingTest {
> -    final toolbox.ToolBox tb = new ToolBox();
> -    final Path base = Paths.get(".");
> -    final String processedSource = """
> -        import module java.base;
> -        import java.lang.annotation.*;
> -        public class Main {
> -          public static void main(String[] args) {
> -            List.of();
> -          }
> -          @Ann
> -          private void test() {}
> -          @Retention(RetentionPolicy.RUNTIME)
> -          @Target(ElementType.METHOD)
> -          public @interface Ann {}
> -        }
> -        """;
> +import toolbox.TestRunner;
> +import toolbox.ToolBox;
> +import toolbox.JavacTask;
> +import toolbox.Task;
> +
> +public class ModuleImportProcessingTest extends TestRunner {
> +    ToolBox tb = new ToolBox();
> +
> +    public ModuleImportProcessingTest() {
> +        super(System.err);
> +    }
> +
> +    protected void runTests() throws Exception {
> +        runTests(m -> new Object[] { Paths.get(m.getName()) });
> +    }
>  
> -    public static void main(String[] args) throws Exception {
> -        new ModuleImportProcessingTest().test();
> +    Path[] findJavaFiles(Path... paths) throws Exception {
> +        return tb.findJavaFiles(paths);
>      }
>  
> -    public void test() throws Exception {
> -        tb.writeJavaFiles(base, processedSource);
> +    public static void main(String... args) throws Exception {
> +        new ModuleImportProcessingTest().runTests();
> +    }
> +
> +    @Test
> +    public void test(Path base) throws Exception {
> +        Path src = base.resolve("src");
> +        tb.writeJavaFiles(src,
> +                """
> +                import module java.base;
> +                import java.lang.annotation.*;
> +                public class Main {
> +                    public static void main(String[] args) {
> +                        List.of();
> +                    }
> +                    @Ann
> +                    private void test() {}
> +
> +                    @Retention(RetentionPolicy.RUNTIME)
> +                    @Target(ElementType.METHOD)
> +                    public @interface Ann {}
> +                }
> +                """);
>          new toolbox.JavacTask(tb)
>                  .options(
>                          "-processor", AP.class.getName(),
> @@ -73,24 +91,15 @@ public void test() throws Exception {
>                          "-source", Integer.toString(Runtime.version().feature()),
>                          "-proc:only"
>                  )
> -                .outdir(base.toString())
> -                .files(base.resolve("Main.java"))
> -                .run(Task.Expect.SUCCESS)
> -                .writeAll();
> +                .files(findJavaFiles(src))
> +                .run();
>      }
>  
>      @SupportedAnnotationTypes("*")
>      public static final class AP extends AbstractProcessor {
> -
>          @Override
>          public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
>              return false;
>          }
> -
> -        @Override
> -        public SourceVersion getSupportedSourceVersion() {
> -            return SourceVersion.latest();
> -        }
> -
>      }
> -}
> \ No newline at end of file
> +}
> ```

Thanks for suggestions. I have sliglty changed the test code you have provided. I saw that people use base.resolve("src") for source file path, but as I understand, the context of base is flushed after each test so effectively it doesnt make much of difference but adds some noise in the code making harder to distinguish the flow of test. Same goes for test source string. I thought that this way test may be more readable.

As for TreeTranslator, I also have thought about adding visit method in this issue, but I figured that it doesnt really relate to the issue title so this change could kind of subtle. Still, If you think it should be fixed in the same issue, than I dont see anything against it as long as such expromt changes are usual there. Maybe its worth renaming the issue to something like "Some visitors dont override visitModuleImport resulting in AssertionError" though.

I also guess maybe if we add visitModuleImport for TreeTranslator here then we should also add a test for it just in case, but I didnt find mockito in jtreg suite so im not sure that it is possible to verify how many times exactly method has been called and writing full translation test just to verify simple 2-line methods works correctly seems like overkill. (I didnt even find any tests for TreeTranslator, seems like its not covered by tests at all)

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (liach) [inline test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:60]:
```suggestion
            public static void main(String[] args) {
                List.of();
            }
            @Ann
            private void test() {}
            @Retention(RetentionPolicy.RUNTIME)
            @Target(ElementType.METHOD)
            public @interface Ann {}
```
Indentation.

File: test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
Comment (liach) [inline test/langtools/tools/javac/processing/ModuleImportProcessingTest.java:73]:
```suggestion
        runTests(m -> new Object[] { Path.of(m.getName()) });
```
`Paths` exists because Path was added in JDK 7, before static methods were supported in interfaces in JDK 8.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Comment by vicente-romero-oracle** [comment]
> > ok this is what I would do, please feel free to use it, it is applicable on top of your current code. As you can see I have reformatted the test to align it better with the rest of similar tests in our code base. Also as Jan suggested I have added an equivalent method to TreeTranslator for completeness.
> > ```
> > diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
> > index 59a7457e6d0..63778fb42ff 100644
> > --- a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
> > +++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeTranslator.java
> > @@ -131,6 +131,11 @@ public void visitImport(JCImport tree) {
> >          result = tree;
> >      }
> >  
> > +    public void visitModuleImport(JCModuleImport tree) {
> > +        tree.module = translate(tree.module);
> > +        result = tree;
> > +    }
> > +
> >      public void visitClassDef(JCClassDecl tree) {
> >          tree.mods = translate(tree.mods);
> >          tree.typarams = translateTypeParams(tree.typarams);
> > diff --git a/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java b/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
> > index 88c7974a26e..760ec30820e 100644
> > --- a/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
> > +++ b/test/langtools/tools/javac/processing/ModuleImportProcessingTest.java
> > @@ -21,7 +21,16 @@
> >   * questions.
> >   */
> >  
> > -import toolbox.*;
> > +/**
> > + * @test
> > + * @bug 8332497
> > + * @summary javac prints an AssertionError when annotation processing runs on program with module imports
> > + * @library /tools/lib
> > + * @modules jdk.compiler/com.sun.tools.javac.api
> > + *          jdk.compiler/com.sun.tools.javac.main
> > + * @build toolbox.JavacTask toolbox.ToolBox toolbox.Task
> > + * @run main ModuleImportProcessingTest
> > + */
> >  
> >  import javax.annotation.processing.AbstractProcessor;
> >  import javax.annotation.processing.RoundEnvironment;
> > @@ -32,40 +41,49 @@
> >  import java.nio.file.Paths;
> >  import java.util.Set;
> >  
> > -/**
> > - * @test
> > - * @bug 8332497
> > - * @summary error: javac prints an AssertionError when annotation processing runs on program with module imports
> > - * @library /tools/lib
> > - * @modules jdk.compiler/com.sun.tools.javac.api
> > - *          jdk.compiler/com.sun.tools.javac.main
> > - * @build toolbox.JavacTask toolbox.ToolBox toolbox.Task
> > - * @run main ModuleImportProcessingTest
> > - */
> > -public class ModuleImportProcessingTest {
> > -    final toolbox.ToolBox tb = new ToolBox();
> > -    final Path base = Paths.get(".");
> > -    final String processedSource = """
> > -        import module java.base;
> > -        import java.lang.annotation.*;
> > -        public class Main {
> > -          public static void main(String[] args) {
> > -            List.of();
> > -          }
> > -          @Ann
> > -          private void test() {}
> > -          @Retention(RetentionPolicy.RUNTIME)
> > -          @Target(ElementType.METHOD)
> > -          public @interface Ann {}
> > -        }
> > -        """;
> > +import toolbox.TestRunner;
> > +import toolbox.ToolBox;
> > +import toolbox.JavacTask;
> > +import toolbox.Task;
> > +
> > +public class ModuleImportProcessingTest extends TestRunner {
> > +    ToolBox tb = new ToolBox();
> > +
> > +    public ModuleImportProcessingTest() {
> > +        super(System.err);
> > +    }
> > +
> > +    protected void runTests() throws Exception {
> > +        runTests(m -> new Object[] { Paths.get(m.getName()) });
> > +    }
> >  
> > -    public static void main(String[] args) throws Exception {
> > -        new ModuleImportProcessingTest().test();
> > +    Path[] findJavaFiles(Path... paths) throws Exception {
> > +        return tb.findJavaFiles(paths);
> >      }
> >  
> > -    public void test() throws Exception {
> > -        tb.writeJavaFiles(base, processedSource);
> > +    public static void main(String... args) throws Exception {
> > +        new ModuleImportProcessingTest().runTests();
> > +    }
> > +
> > +    @Test
> > +    public void test(Path base) throws Exception {
> > +        Path src = base.resolve("src");
> > +        tb.writeJavaFiles(src,
> > +                """
> > +                import module java.base;
> > +                import java.lang.annotation.*;
> > +                public class Main {
> > +                    public static void main(String[] args) {
> > +                        List.of();
> > +                    }
> > +                    @Ann
> > +                    private void test() {}
> > +
> > +                    @Retention(RetentionPolicy.RUNTIME)
> > +                    @Target(ElementType.METHOD)
> > +                    public @interface Ann {}
> > +                }
> > +                """);
> >          new toolbox.JavacTask(tb)
> >                  .options(
> >                          "-processor", AP.class.getName(),
> > @@ -73,24 +91,15 @@ public void test() throws Exception {
> >                          "-source", Integer.toString(Runtime.version().feature()),
> >                          "-proc:only"
> >                  )
> > -                .outdir(base.toString())
> > -                .files(base.resolve("Main.java"))
> > -                .run(Task.Expect.SUCCESS)
> > -                .writeAll();
> > +                .files(findJavaFiles(src))
> > +                .run();
> >      }
> >  
> >      @SupportedAnnotationTypes("*")
> >      public static final class AP extends AbstractProcessor {
> > -
> >          @Override
> >          public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
> >              return false;
> >          }
> > -
> > -        @Override
> > -        public SourceVersion getSupportedSourceVersion() {
> > -            return SourceVersion.latest();
> > -        }
> > -
> >      }
> > -}
> > \ No newline at end of file
> > +}
> > ```
> 
> Thanks for suggestions. I have sliglty changed the test code you have provided. I saw that people use base.resolve("src") for source file path, but as I understand, the context of base is flushed after each test so effectively it doesnt make much of a difference but adds some noise in the code making harder to distinguish the flow of test. Same goes for test source string. I thought that this way test may be more readable.
> 
> As for TreeTranslator, I also have thought about adding visit method in this issue, but I figured that it doesnt really relate to the issue title so this change could kind of subtle. Still, If you think it should be fixed in the same issue, than I dont see anything against it as long as such expromt changes are usual there. Maybe its worth renaming the issue to something like "Some visitors dont override visitModuleImport resulting in AssertionError" though.
> 
> I also guess maybe if we add visitModuleImport for TreeTranslator here then we should also add a test for it just in case, but I didnt find mockito in jtreg suite so im not sure that it is possible to verify how many times exactly method has been called and writing full translation test just to verify simple 2-line methods works correctly seems like overkill. (I didnt even find any tests for TreeTranslator, seems like its not covered by tests at all)

yes we don't have tests for TreeTranslator, at least not targeting it explicitly, so it is OK not to provide one now. I think the title of the bug is OK as it is

**Review by liach — APPROVED** [review]

**Comment by Evemose** [comment]
@vicente-romero-oracle could you sponsor please?

**Review by lahodaj — APPROVED** [review]
Looks good to me. Thanks for working on this!

I am running tests now.

@Evemose, please note you'll need to type:
```
/integrate
```
in a separate comment, so that a committer can sponsor.

Thanks!

**Comment by lahodaj** [comment]
> Hi! I have already typed integrate a bit earlier in the conversation ([#19292 (comment)](https://github.com/openjdk/jdk/pull/19292#issuecomment-2131213489)), so I am waiting for sponsor for now

Right. Sorry, I missed the label.

**Pull request closed without merging**
